From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Peter=20G=C3=BCttinger?= <njol@gmx.ch>
Date: Wed, 7 Sep 2022 08:28:57 +0200
Subject: [PATCH] Monumenta - add IFRAMES DamageModifier


diff --git a/src/main/java/org/bukkit/event/entity/EntityDamageEvent.java b/src/main/java/org/bukkit/event/entity/EntityDamageEvent.java
index 6241f31e7326954b2e704da46bf0115adbcf8671..96aef06a479cd39b6e55206043481d4c0184a6cf 100644
--- a/src/main/java/org/bukkit/event/entity/EntityDamageEvent.java
+++ b/src/main/java/org/bukkit/event/entity/EntityDamageEvent.java
@@ -218,6 +218,11 @@ public class EntityDamageEvent extends EntityEvent implements Cancellable {
          * raw {@link EntityDamageEvent#getDamage()}.
          */
         BASE,
+        // Monumenta: added this
+        /**
+         * This represents the damage reduction caused by active iframes.
+         */
+        IFRAMES,
         /**
          * This represents the damage reduced by a wearing a helmet when hit
          * by a falling block.
