From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tim Southwick <NickNackGus@gmail.com>
Date: Sun, 19 Apr 2020 19:34:47 -0400
Subject: [PATCH] Monumenta - Selectors require targets to be alive


diff --git a/src/main/java/net/minecraft/commands/arguments/selector/EntitySelectorParser.java b/src/main/java/net/minecraft/commands/arguments/selector/EntitySelectorParser.java
index 4cc2ee22369aef22c9bf9d2b244c7278d6db39bb..92d1eccf0c07e7b4b6368a403543f07acabf14e0 100644
--- a/src/main/java/net/minecraft/commands/arguments/selector/EntitySelectorParser.java
+++ b/src/main/java/net/minecraft/commands/arguments/selector/EntitySelectorParser.java
@@ -130,9 +130,7 @@ public class EntitySelectorParser {
         this.level = MinMaxBounds.Ints.ANY;
         this.rotX = WrappedMinMaxBounds.ANY;
         this.rotY = WrappedMinMaxBounds.ANY;
-        this.predicate = (entity) -> {
-            return true;
-        };
+        this.predicate = Entity::isAlive;
         this.order = EntitySelector.ORDER_ARBITRARY;
         this.suggestions = EntitySelectorParser.SUGGEST_NOTHING;
         this.reader = reader;
