From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Byron Marohn <combustible@live.com>
Date: Thu, 16 Apr 2020 19:50:43 -0400
Subject: [PATCH] Monumenta - Max level for curse of vanishing is 2

Signed-off-by: Byron Marohn <combustible@live.com>

diff --git a/src/main/java/net/minecraft/world/item/enchantment/VanishingCurseEnchantment.java b/src/main/java/net/minecraft/world/item/enchantment/VanishingCurseEnchantment.java
index 293ba187d0794cbdb26d4706143219b680cd64be..1b41065dd776a58dd907a2e526642876c6fe7fb5 100644
--- a/src/main/java/net/minecraft/world/item/enchantment/VanishingCurseEnchantment.java
+++ b/src/main/java/net/minecraft/world/item/enchantment/VanishingCurseEnchantment.java
@@ -17,6 +17,11 @@ public class VanishingCurseEnchantment extends Enchantment {
         return 50;
     }
 
+    @Override
+    public int getMaxLevel() {
+        return 2;
+    }
+
     @Override
     public boolean isTreasureOnly() {
         return true;
