From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Byron Marohn <combustible@live.com>
Date: Fri, 6 Sep 2019 20:36:55 -0400
Subject: [PATCH] Monumenta - DamageCause.CUSTOM support for class abilities

Signed-off-by: Byron Marohn <combustible@live.com>

diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 221f5088953b3452966d07eabd4ea8b38c465fd9..60da9faad634816516ea2abc4f3c967d414ff254 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -1025,6 +1025,8 @@ public class CraftEventFactory {
                 } else if (damager.getBukkitEntity() instanceof Projectile) {
                     cause = DamageCause.PROJECTILE;
                 }
+            } else if ("custom".equals(source.getMsgId())) {
+                cause = DamageCause.CUSTOM;
             } else if (source.is(DamageTypes.THORNS)) {
                 cause = DamageCause.THORNS;
             } else if (source.is(DamageTypes.SONIC_BOOM)) {
