From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Byron Marohn <combustible@live.com>
Date: Fri, 6 Sep 2019 20:36:55 -0400
Subject: [PATCH] Monumenta - DamageCause.CUSTOM support for class abilities

Signed-off-by: Byron Marohn <combustible@live.com>

diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index a3c8a0291fa9ae6f3c96d937dd4621edd7c48535..d83532c451916106b9ef973b15d3432222efea9e 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -1003,6 +1003,8 @@ public class CraftEventFactory {
                 } else if (damager.getBukkitEntity() instanceof Projectile) {
                     cause = DamageCause.PROJECTILE;
                 }
+            } else if ("custom".equals(source.msgId)) {
+                cause = DamageCause.CUSTOM;
             } else if ("thorns".equals(source.msgId)) {
                 cause = DamageCause.THORNS;
             }
