From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Byron Marohn <combustible@live.com>
Date: Mon, 5 Apr 2021 14:04:44 -0400
Subject: [PATCH] Monumenta - Move spawnpoint set for sleeping in bed after
 event

Signed-off-by: Byron Marohn <combustible@live.com>

diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index f5d4d51e64e80d07eaea0b6cb8cc79ae432f0b23..4994f5a2f0f641aee45b8bcc6eca40cd87798764 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -1329,12 +1329,7 @@ public class ServerPlayer extends Player {
             } else if (this.bedBlocked(blockposition, enumdirection)) {
                 return Either.left(Player.BedSleepingProblem.OBSTRUCTED);
             } else {
-                this.setRespawnPosition(this.level.dimension(), blockposition, this.getYRot(), false, true, com.destroystokyo.paper.event.player.PlayerSetSpawnEvent.Cause.BED); // Paper - PlayerSetSpawnEvent
-                if (this.level.isDay()) {
-                    return Either.left(Player.BedSleepingProblem.NOT_POSSIBLE_NOW);
-                } else {
-                    return Either.right(Unit.INSTANCE);
-                }
+                return Either.right(Unit.INSTANCE);
             }
         } else {
             return Either.left(Player.BedSleepingProblem.OTHER_PROBLEM);
@@ -1359,6 +1354,8 @@ public class ServerPlayer extends Player {
             return bedResult;
         }
 
+        this.setRespawnPosition(this.level.dimension(), blockposition, this.getYRot(), false, true, com.destroystokyo.paper.event.player.PlayerSetSpawnEvent.Cause.BED); // Paper - PlayerSetSpawnEvent
+
         {
             {
                 {
