From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Byron Marohn <combustible@live.com>
Date: Wed, 29 Apr 2020 21:17:15 +0000
Subject: [PATCH] Monumenta - Reset last player hurt time on taking any damage

Signed-off-by: Byron Marohn <combustible@live.com>

diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 95a27d28f73039693ca64601954af62028413634..0d128ddb413abc09dbad036bf64c8863d33101f3 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -1408,6 +1408,10 @@ public abstract class LivingEntity extends Entity implements Attackable {
 
             Entity entity1 = source.getEntity();
 
+            if (this.lastHurtByPlayer != null) {
+                this.lastHurtByPlayerTime = 100;
+            }
+
             if (entity1 != null) {
                 if (entity1 instanceof LivingEntity) {
                     LivingEntity entityliving1 = (LivingEntity) entity1;
