From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Byron Marohn <combustible@live.com>
Date: Wed, 29 Apr 2020 21:17:15 +0000
Subject: [PATCH] Monumenta - Reset last player hurt time on taking any damage

Signed-off-by: Byron Marohn <combustible@live.com>

diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 1521f53ee1bd85ca44a68b2c9d969eaf63fa342e..b9b7ed77fc2ad890c4bfde4b265720fc2623c36a 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -1372,6 +1372,10 @@ public abstract class LivingEntity extends Entity {
             this.hurtDir = 0.0F;
             Entity entity1 = source.getEntity();
 
+            if (this.lastHurtByPlayer != null) {
+                this.lastHurtByPlayerTime = 100;
+            }
+
             if (entity1 != null) {
                 if (entity1 instanceof LivingEntity && !source.isNoAggro()) {
                     this.setLastHurtByMob((LivingEntity) entity1);
