From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joel Ong <ongjiaming2001@gmail.com>
Date: Tue, 24 May 2022 18:58:19 +1000
Subject: [PATCH] Monumenta - Clear Crafting Slots when Clearing


diff --git a/src/main/java/net/minecraft/world/entity/player/Inventory.java b/src/main/java/net/minecraft/world/entity/player/Inventory.java
index 27c028ab6b1edb6e413af3bbaa27bf30f2d85540..842378ff03d36f528ab1a87f86a23dfc9017c06d 100644
--- a/src/main/java/net/minecraft/world/entity/player/Inventory.java
+++ b/src/main/java/net/minecraft/world/entity/player/Inventory.java
@@ -21,6 +21,8 @@ import net.minecraft.world.Nameable;
 import net.minecraft.world.damagesource.DamageSource;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EquipmentSlot;
+import net.minecraft.world.inventory.CraftingMenu;
+import net.minecraft.world.inventory.InventoryMenu;
 import net.minecraft.world.item.ArmorItem;
 import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
@@ -254,6 +256,34 @@ public class Inventory implements Container, Nameable {
             this.player.containerMenu.setCarried(ItemStack.EMPTY);
         }
 
+        // Monumenta Start
+        net.minecraft.world.inventory.CraftingContainer craftingContainer =
+                this.player.containerMenu instanceof net.minecraft.world.inventory.InventoryMenu ? ((InventoryMenu) this.player.containerMenu).craftSlots :
+                this.player.containerMenu instanceof net.minecraft.world.inventory.CraftingMenu ? ((CraftingMenu) this.player.containerMenu).craftSlots : null;
+
+        if (craftingContainer != null) {
+            for (int slotNum = 0; slotNum < craftingContainer.getContainerSize(); ++slotNum) {
+                ItemStack item = craftingContainer.getItem(slotNum);
+
+                if (!item.isEmpty() && shouldRemove.test(item)) {
+                    int countFromStack = maxCount <= 0 ? item.getCount() : Math.min(maxCount - j, item.getCount());
+
+                    j += countFromStack;
+                    if (maxCount != 0) {
+                        item.shrink(countFromStack);
+                        if (item.isEmpty()) {
+                            craftingContainer.setItem(slotNum, ItemStack.EMPTY);
+                        }
+
+                        if (maxCount > 0 && j >= maxCount) {
+                            return j;
+                        }
+                    }
+                }
+            }
+        }
+
+        // Monumenta End
         return j;
     }
 
diff --git a/src/main/java/net/minecraft/world/inventory/InventoryMenu.java b/src/main/java/net/minecraft/world/inventory/InventoryMenu.java
index da0f5c5e6ca7ce7b38792e6da52c5cdcdbae3b78..740bb86321dac9423112e840024ceac86639ad3a 100644
--- a/src/main/java/net/minecraft/world/inventory/InventoryMenu.java
+++ b/src/main/java/net/minecraft/world/inventory/InventoryMenu.java
@@ -39,7 +39,7 @@ public class InventoryMenu extends RecipeBookMenu<CraftingContainer> {
     static final ResourceLocation[] TEXTURE_EMPTY_SLOTS = new ResourceLocation[]{InventoryMenu.EMPTY_ARMOR_SLOT_BOOTS, InventoryMenu.EMPTY_ARMOR_SLOT_LEGGINGS, InventoryMenu.EMPTY_ARMOR_SLOT_CHESTPLATE, InventoryMenu.EMPTY_ARMOR_SLOT_HELMET};
     private static final EquipmentSlot[] SLOT_IDS = new EquipmentSlot[]{EquipmentSlot.HEAD, EquipmentSlot.CHEST, EquipmentSlot.LEGS, EquipmentSlot.FEET};
     // CraftBukkit start
-    private final CraftingContainer craftSlots;
+    public final CraftingContainer craftSlots; // Monumenta
     private final ResultContainer resultSlots;
     // CraftBukkit end
     public final boolean active;
