From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ziggleflig <dmdinnc@gmail.com>
Date: Sun, 1 May 2022 00:46:53 -0400
Subject: [PATCH] Monumenta - Fix shulker NoAI allowing peeking


diff --git a/src/main/java/net/minecraft/world/entity/monster/Shulker.java b/src/main/java/net/minecraft/world/entity/monster/Shulker.java
index 8cb910da17d75a9d9c7dbeb3c9e24b6de657a2f7..542babb33b9f79d15bd63d2c364f59b239e59c26 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Shulker.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Shulker.java
@@ -87,6 +87,8 @@ public class Shulker extends AbstractGolem implements VariantHolder<Optional<Dye
     });
     private float currentPeekAmountO;
     private float currentPeekAmount;
+    private float initialPeekAmount;
+    private float initialPeekAmount0;
     @Nullable
     private BlockPos clientOldAttachPosition;
     private int clientSideTeleportInterpolation;
@@ -94,6 +96,8 @@ public class Shulker extends AbstractGolem implements VariantHolder<Optional<Dye
 
     public Shulker(EntityType<? extends Shulker> type, Level world) {
         super(type, world);
+        this.initialPeekAmount = this.currentPeekAmount;
+        this.initialPeekAmount0 = this.currentPeekAmountO;
         this.xpReward = 5;
         this.lookControl = new Shulker.ShulkerLookControl(this);
     }
@@ -181,7 +185,7 @@ public class Shulker extends AbstractGolem implements VariantHolder<Optional<Dye
     @Override
     public void tick() {
         super.tick();
-        if (!this.level.isClientSide && !this.isPassenger() && !this.canStayAt(this.blockPosition(), this.getAttachFace())) {
+        if (!this.level.isClientSide && !this.isPassenger() && !this.canStayAt(this.blockPosition(), this.getAttachFace()) && !this.isNoAi()) {
             this.findNewAttachment();
         }
 
@@ -241,6 +245,10 @@ public class Shulker extends AbstractGolem implements VariantHolder<Optional<Dye
     }
 
     private void onPeekAmountChange() {
+        if (this.isNoAi()) {
+            this.currentPeekAmount = this.initialPeekAmount;
+            this.currentPeekAmountO = this.initialPeekAmount0;
+        }
         this.reapplyPosition();
         float f = Shulker.getPhysicalPeek(this.currentPeekAmount);
         float f1 = Shulker.getPhysicalPeek(this.currentPeekAmountO);
