From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ziggleflig <dmdinnc@gmail.com>
Date: Sun, 1 May 2022 00:46:53 -0400
Subject: [PATCH] Monumenta - Fix shulker NoAI allowing peeking


diff --git a/src/main/java/net/minecraft/world/entity/monster/Shulker.java b/src/main/java/net/minecraft/world/entity/monster/Shulker.java
index a9dfe190f46230077e2e1bf9aacbf5375651f216..4992f088cc71b1b30df5f5b0d205f4d174bfe310 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Shulker.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Shulker.java
@@ -81,6 +81,8 @@ public class Shulker extends AbstractGolem implements Enemy {
     });
     private float currentPeekAmountO;
     private float currentPeekAmount;
+    private float initialPeekAmount;
+    private float initialPeekAmount0;
     @Nullable
     private BlockPos clientOldAttachPosition;
     private int clientSideTeleportInterpolation;
@@ -88,6 +90,8 @@ public class Shulker extends AbstractGolem implements Enemy {
 
     public Shulker(EntityType<? extends Shulker> type, Level world) {
         super(type, world);
+        this.initialPeekAmount = this.currentPeekAmount;
+        this.initialPeekAmount0 = this.currentPeekAmountO;
         this.xpReward = 5;
         this.lookControl = new Shulker.ShulkerLookControl(this);
     }
@@ -175,7 +179,7 @@ public class Shulker extends AbstractGolem implements Enemy {
     @Override
     public void tick() {
         super.tick();
-        if (!this.level.isClientSide && !this.isPassenger() && !this.canStayAt(this.blockPosition(), this.getAttachFace())) {
+        if (!this.level.isClientSide && !this.isPassenger() && !this.canStayAt(this.blockPosition(), this.getAttachFace()) && !this.isNoAi()) {
             this.findNewAttachment();
         }
 
@@ -235,6 +239,10 @@ public class Shulker extends AbstractGolem implements Enemy {
     }
 
     private void onPeekAmountChange() {
+        if (this.isNoAi()) {
+            this.currentPeekAmount = this.initialPeekAmount;
+            this.currentPeekAmountO = this.initialPeekAmount0;
+        }
         this.reapplyPosition();
         float f = Shulker.getPhysicalPeek(this.currentPeekAmount);
         float f1 = Shulker.getPhysicalPeek(this.currentPeekAmountO);
