From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joel Ong <ongjiaming2001@gmail.com>
Date: Wed, 25 May 2022 12:44:41 +1000
Subject: [PATCH] Monumenta - Fix Generating Loot Table Contents


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftLootTable.java b/src/main/java/org/bukkit/craftbukkit/CraftLootTable.java
index 46a7585d1a161dd598e2cdd7244874390fe821c5..f147acf1f44f44fb053b0f08e41985f5083a4dc7 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftLootTable.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftLootTable.java
@@ -86,7 +86,7 @@ public class CraftLootTable implements org.bukkit.loot.LootTable {
         }
         this.setMaybe(builder, LootContextParams.ORIGIN, new Vec3(loc.getX(), loc.getY(), loc.getZ()));
         if (this.getHandle() != LootTable.EMPTY) {
-            // builder.luck(context.getLuck());
+            builder.withLuck(context.getLuck());
 
             if (context.getLootedEntity() != null) {
                 Entity nmsLootedEntity = ((CraftEntity) context.getLootedEntity()).getHandle();
@@ -113,12 +113,10 @@ public class CraftLootTable implements org.bukkit.loot.LootTable {
         // SPIGOT-5603 - Avoid IllegalArgumentException in LootTableInfo#build()
         LootContextParamSet.Builder nmsBuilder = new LootContextParamSet.Builder();
         for (LootContextParam<?> param : this.getHandle().getParamSet().getRequired()) {
-            nmsBuilder.required(param);
+            nmsBuilder.optional(param); // Monumenta
         }
         for (LootContextParam<?> param : this.getHandle().getParamSet().getAllowed()) {
-            if (!this.getHandle().getParamSet().getRequired().contains(param)) {
-                nmsBuilder.optional(param);
-            }
+            nmsBuilder.optional(param); // Monumenta
         }
         nmsBuilder.optional(LootContextParams.LOOTING_MOD);
 
