From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joel Ong <ongjiaming2001@gmail.com>
Date: Wed, 25 May 2022 12:44:41 +1000
Subject: [PATCH] Monumenta - Fix Generating Loot Table Contents


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftLootTable.java b/src/main/java/org/bukkit/craftbukkit/CraftLootTable.java
index ecbce397f06fcdaa4e72e49cc9765d9736abdd00..9b851f39abf4e3073a199d376b3cee19bc483756 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftLootTable.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftLootTable.java
@@ -87,7 +87,7 @@ public class CraftLootTable implements org.bukkit.loot.LootTable {
         }
         this.setMaybe(builder, LootContextParams.ORIGIN, CraftLocation.toVec3D(loc));
         if (this.getHandle() != LootTable.EMPTY) {
-            // builder.luck(context.getLuck());
+            builder.withLuck(context.getLuck()); // Monumenta
 
             if (context.getLootedEntity() != null) {
                 Entity nmsLootedEntity = ((CraftEntity) context.getLootedEntity()).getHandle();
@@ -114,12 +114,10 @@ public class CraftLootTable implements org.bukkit.loot.LootTable {
         // SPIGOT-5603 - Avoid IllegalArgumentException in LootTableInfo#build()
         LootContextParamSet.Builder nmsBuilder = new LootContextParamSet.Builder();
         for (LootContextParam<?> param : this.getHandle().getParamSet().getRequired()) {
-            nmsBuilder.required(param);
+            nmsBuilder.optional(param); // Monumenta
         }
         for (LootContextParam<?> param : this.getHandle().getParamSet().getAllowed()) {
-            if (!this.getHandle().getParamSet().getRequired().contains(param)) {
-                nmsBuilder.optional(param);
-            }
+            nmsBuilder.optional(param); // Monumenta
         }
         nmsBuilder.optional(LootContextParams.LOOTING_MOD);
 
