From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joel Ong <ongjiaming2001@gmail.com>
Date: Mon, 20 Jun 2022 21:30:32 +1000
Subject: [PATCH] Monumenta - Disable mob dismounting underwater


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 6c98e24f26e06b08ff5c705f61da58724e8615d7..61109566c2147f07dac9a5956eda1b68381c42e9 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -438,9 +438,10 @@ public abstract class LivingEntity extends Entity {
                     }
                 }
 
+                /* MONUMENTA - Stop mobs from leaving "vehicle" when underwater.
                 if (!this.level.isClientSide && this.isPassenger() && this.getVehicle() != null && !this.getVehicle().rideableUnderWater()) {
                     this.stopRiding();
-                }
+                } */
             } else if (this.getAirSupply() < this.getMaxAirSupply()) {
                 this.setAirSupply(this.increaseAirSupply(this.getAirSupply()));
             }
