From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joel Ong <ongjiaming2001@gmail.com>
Date: Wed, 25 May 2022 14:23:42 +1000
Subject: [PATCH] Monumenta - Mobs that despawn return to their spawners


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 4994f5a2f0f641aee45b8bcc6eca40cd87798764..03eeee5b0adf1da67c6848544de47c62eeba831c 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -989,6 +989,15 @@ public class ServerPlayer extends Player {
         this.setSharedFlagOnFire(false);
         this.getCombatTracker().recheckStatus();
         this.setLastDeathLocation(Optional.of(GlobalPos.of(this.level.dimension(), this.blockPosition())));
+
+        // Monumenta START - don't allow mobs to reprime spawners when they are near players that die.
+        for (org.bukkit.entity.Entity nearby : this.level.getWorld().getNearbyEntities(new Location(this.level.getWorld(), this.getX(), this.getY(), this.getZ()), 24.0d, 24.0d, 24.0d)) {
+            ((org.bukkit.craftbukkit.entity.CraftEntity)nearby).getHandle().spawnerSpawnedBy = null;
+            if (this.getTags().contains("DelvesPlayer")) {
+                ((org.bukkit.craftbukkit.entity.CraftEntity)nearby).getHandle().delveReprime = true;
+            }
+        }
+        // Monumenta END
     }
 
     private void tellNeutralMobsThatIDied() {
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 7555b04dcf274bb624b89f2eb9ff80da0056de4e..ad2b3d67645497d8162b723a47d9917c2d640d1c 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -392,6 +392,8 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     public boolean fromNetherPortal; // Paper
     protected int numCollisions = 0; // Paper
     public boolean spawnedViaMobSpawner; // Paper - Yes this name is similar to above, upstream took the better one
+    public net.minecraft.world.level.BaseSpawner spawnerSpawnedBy = null; // Monumenta
+    public boolean delveReprime = false; // Monumenta
     @javax.annotation.Nullable
     private org.bukkit.util.Vector origin;
     @javax.annotation.Nullable
@@ -673,6 +675,33 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
         return this.id;
     }
 
+    // Monumenta START
+    public void despawn() {
+        if (this instanceof net.minecraft.world.entity.LivingEntity && this.spawnerSpawnedBy != null && ((net.minecraft.world.entity.LivingEntity) this).getHealth() >= 1 && this.getY() > 0 && this.spawnerSpawnedBy.blockPos != null) {
+            // Get closest player to spawner
+            Player player = this.level.getNearestPlayer(this.spawnerSpawnedBy.blockPos.getX(), this.spawnerSpawnedBy.blockPos.getY(), this.spawnerSpawnedBy.blockPos.getZ(), -1, false);
+            if (player != null) {
+                // Figure out how far player is from spawner.
+                double dX = player.getX() - this.spawnerSpawnedBy.blockPos.getX();
+                double dY = player.getY() - this.spawnerSpawnedBy.blockPos.getY();
+                double dZ = player.getZ() - this.spawnerSpawnedBy.blockPos.getZ();
+                double nearestPlayerDistanceSquared = dX * dX + dY * dY + dZ * dZ;
+                if (nearestPlayerDistanceSquared >= 576) { // 24 * 24
+                    // No players are next to the spawner the mob came from - reprime it
+                    if (this.delveReprime) {
+                        // Player died nearby in a delve -> reprime to 5s
+                        this.spawnerSpawnedBy.spawnDelay = 100;
+                    } else {
+                        // Not a delve or no player died nearby -> fully reprime
+                        this.spawnerSpawnedBy.spawnDelay = 0;
+                    }
+                }
+            }
+        }
+        discard();
+    }
+    // Monumenta END
+
     public void remove(Entity.RemovalReason reason) {
         this.setRemoved(reason);
     }
diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index 49b983064ea810382b6112f5dc7f93ba4e5710bd..a219e45fb1a8709cbab7839cd5535bc2d108d17f 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -824,7 +824,7 @@ public abstract class Mob extends LivingEntity {
     @Override
     public void checkDespawn() {
         if (this.level.getDifficulty() == Difficulty.PEACEFUL && this.shouldDespawnInPeaceful()) {
-            this.discard();
+            this.despawn(); // Monumenta
         } else if (!this.isPersistenceRequired() && !this.requiresCustomPersistence()) {
             // Paper start - optimise checkDespawn
             Player entityhuman = this.level.findNearbyPlayer(this, level.paperConfig().entities.spawning.despawnRanges.get(this.getType().getCategory()).hard() + 1, EntitySelector.PLAYER_AFFECTS_SPAWNING); // Paper
@@ -839,14 +839,14 @@ public abstract class Mob extends LivingEntity {
                 int j = i * i;
 
                 if (d0 > (double) j && this.removeWhenFarAway(d0)) {
-                    this.discard();
+                    this.despawn(); // Monumenta
                 }
 
                 int k = this.level.paperConfig().entities.spawning.despawnRanges.get(this.getType().getCategory()).soft(); // Paper - custom despawn distances
                 int l = k * k;
 
                 if (this.noActionTime > 600 && this.random.nextInt(800) == 0 && d0 > (double) l && this.removeWhenFarAway(d0)) {
-                    this.discard();
+                    this.despawn(); // Monumenta
                 } else if (d0 < (double) l) {
                     this.noActionTime = 0;
                 }
diff --git a/src/main/java/net/minecraft/world/level/BaseSpawner.java b/src/main/java/net/minecraft/world/level/BaseSpawner.java
index 1b448ea7b4de036acfb9db53a88042c92921adf5..694765318c773227edfef62d4bdb34adf898eb47 100644
--- a/src/main/java/net/minecraft/world/level/BaseSpawner.java
+++ b/src/main/java/net/minecraft/world/level/BaseSpawner.java
@@ -46,6 +46,7 @@ public abstract class BaseSpawner {
     public int requiredPlayerRange = 16;
     public int spawnRange = 4;
     private int tickDelay = 0; // Paper
+    public BlockPos blockPos = null; // Monumenta
 
     public BaseSpawner() {}
 
@@ -187,6 +188,12 @@ public abstract class BaseSpawner {
                             // Spigot End
                         }
                         entity.spawnedViaMobSpawner = true; // Paper
+                        // Monumenta START
+                        if (!(entity instanceof net.minecraft.world.entity.FlyingMob || entity instanceof net.minecraft.world.entity.monster.Vex)) {
+                            entity.spawnerSpawnedBy = this;
+                        }
+                        // Monumenta END
+
                         entity.spawnReason = org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SPAWNER; // Paper
                         flag = true; // Paper
                         // Spigot Start
diff --git a/src/main/java/net/minecraft/world/level/block/entity/SpawnerBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/SpawnerBlockEntity.java
index 46a6f604ea14fe29d8a5f12dda4e533a4eb61637..5e49d7bc8ee06080a74a387b26eba70b849d85d1 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/SpawnerBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/SpawnerBlockEntity.java
@@ -33,6 +33,7 @@ public class SpawnerBlockEntity extends BlockEntity {
 
     public SpawnerBlockEntity(BlockPos pos, BlockState state) {
         super(BlockEntityType.MOB_SPAWNER, pos, state);
+        this.spawner.blockPos = pos; // Monumenta
     }
 
     @Override
