From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joel Ong <ongjiaming2001@gmail.com>
Date: Mon, 20 Jun 2022 19:33:37 +1000
Subject: [PATCH] Monumenta - Remove bee death + neutral after sting


diff --git a/src/main/java/net/minecraft/world/entity/animal/Bee.java b/src/main/java/net/minecraft/world/entity/animal/Bee.java
index c33e5c51839c8e6ec04c1b302127d2bf0f48664c..dd4daca91b82c375488e4a30f625e78755b724e3 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Bee.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Bee.java
@@ -262,8 +262,9 @@ public class Bee extends Animal implements NeutralMob, FlyingAnimal {
                 }
             }
 
-            this.setHasStung(true);
-            this.stopBeingAngry();
+            // MONUMENTA - Remove Bee dying + neutral after sting
+            // this.setHasStung(true);
+            // this.stopBeingAngry();
             this.playSound(SoundEvents.BEE_STING, 1.0F, 1.0F);
         }
 
