From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joel Ong <ongjiaming2001@gmail.com>
Date: Mon, 20 Jun 2022 21:30:32 +1000
Subject: [PATCH] Monumenta - Disable mob dismounting underwater


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index b9b7ed77fc2ad890c4bfde4b265720fc2623c36a..6933d267ada333dcddc68c78cc99bb7deab903e5 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -432,9 +432,10 @@ public abstract class LivingEntity extends Entity {
                     }
                 }
 
+                /* MONUMENTA - Stop mobs from leaving "vehicle" when underwater.
                 if (!this.level.isClientSide && this.isPassenger() && this.getVehicle() != null && !this.getVehicle().rideableUnderWater()) {
                     this.stopRiding();
-                }
+                } */
             } else if (this.getAirSupply() < this.getMaxAirSupply()) {
                 this.setAirSupply(this.increaseAirSupply(this.getAirSupply()));
             }
