From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joel Ong <ongjiaming2001@gmail.com>
Date: Wed, 25 May 2022 14:23:42 +1000
Subject: [PATCH] Monumenta - Mobs that despawn return to their spawners


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 23f7a2e69a64b8772d432b3f37cddc376c6f5cec..2bf67659e4701c65d741ebdd4ac1b5fa8ece51a7 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -998,6 +998,15 @@ public class ServerPlayer extends Player {
         this.setSharedFlagOnFire(false);
         this.getCombatTracker().recheckStatus();
         this.setLastDeathLocation(Optional.of(GlobalPos.of(this.level.dimension(), this.blockPosition())));
+
+        // Monumenta START - don't allow mobs to reprime spawners when they are near players that die.
+        for (org.bukkit.entity.Entity nearby : this.level.getWorld().getNearbyEntities(new Location(this.level.getWorld(), this.getX(), this.getY(), this.getZ()), 24.0d, 24.0d, 24.0d)) {
+            ((org.bukkit.craftbukkit.entity.CraftEntity)nearby).getHandle().spawnerSpawnedBy = null;
+            if (this.getTags().contains("DelvesPlayer")) {
+                ((org.bukkit.craftbukkit.entity.CraftEntity)nearby).getHandle().delveReprime = true;
+            }
+        }
+        // Monumenta END
     }
 
     private void tellNeutralMobsThatIDied() {
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 280ee1838106201f5e3ba7753caced6d030f7e55..2a83c11c059e9f2828f5da49b5f60590edc7e448 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -395,6 +395,8 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     public boolean fromNetherPortal; // Paper
     protected int numCollisions = 0; // Paper
     public boolean spawnedViaMobSpawner; // Paper - Yes this name is similar to above, upstream took the better one
+    public net.minecraft.world.level.BaseSpawner spawnerSpawnedBy = null; // Monumenta
+    public boolean delveReprime = false; // Monumenta
     @javax.annotation.Nullable
     private org.bukkit.util.Vector origin;
     @javax.annotation.Nullable
@@ -676,6 +678,33 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
         return this.id;
     }
 
+    // Monumenta START
+    public void despawn() {
+        if (this instanceof net.minecraft.world.entity.LivingEntity && this.spawnerSpawnedBy != null && ((net.minecraft.world.entity.LivingEntity) this).getHealth() >= 1 && this.getY() > 0 && this.spawnerSpawnedBy.blockPos != null) {
+            // Get closest player to spawner
+            Player player = this.level.getNearestPlayer(this.spawnerSpawnedBy.blockPos.getX(), this.spawnerSpawnedBy.blockPos.getY(), this.spawnerSpawnedBy.blockPos.getZ(), -1, false);
+            if (player != null) {
+                // Figure out how far player is from spawner.
+                double dX = player.getX() - this.spawnerSpawnedBy.blockPos.getX();
+                double dY = player.getY() - this.spawnerSpawnedBy.blockPos.getY();
+                double dZ = player.getZ() - this.spawnerSpawnedBy.blockPos.getZ();
+                double nearestPlayerDistanceSquared = dX * dX + dY * dY + dZ * dZ;
+                if (nearestPlayerDistanceSquared >= 576) { // 24 * 24
+                    // No players are next to the spawner the mob came from - reprime it
+                    if (this.delveReprime) {
+                        // Player died nearby in a delve -> reprime to 5s
+                        this.spawnerSpawnedBy.spawnDelay = 100;
+                    } else {
+                        // Not a delve or no player died nearby -> fully reprime
+                        this.spawnerSpawnedBy.spawnDelay = 0;
+                    }
+                }
+            }
+        }
+        discard();
+    }
+    // Monumenta END
+
     public void remove(Entity.RemovalReason reason) {
         this.setRemoved(reason);
     }
diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index 02cb6b8c1d59855ff4a8aad3024fe12007eca0ee..deaef94ff1e631610c11961b8324e0fa720000ec 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -853,7 +853,7 @@ public abstract class Mob extends LivingEntity implements Targeting {
     @Override
     public void checkDespawn() {
         if (this.level.getDifficulty() == Difficulty.PEACEFUL && this.shouldDespawnInPeaceful()) {
-            this.discard();
+            this.despawn(); // Monumenta
         } else if (!this.isPersistenceRequired() && !this.requiresCustomPersistence()) {
             // Paper start - optimise checkDespawn
             Player entityhuman = this.level.findNearbyPlayer(this, level.paperConfig().entities.spawning.despawnRanges.get(this.getType().getCategory()).hard() + 1, EntitySelector.PLAYER_AFFECTS_SPAWNING); // Paper
@@ -868,14 +868,14 @@ public abstract class Mob extends LivingEntity implements Targeting {
                 int j = i * i;
 
                 if (d0 > (double) j && this.removeWhenFarAway(d0)) {
-                    this.discard();
+                    this.despawn(); // Monumenta
                 }
 
                 int k = this.level.paperConfig().entities.spawning.despawnRanges.get(this.getType().getCategory()).soft(); // Paper - custom despawn distances
                 int l = k * k;
 
                 if (this.noActionTime > 600 && this.random.nextInt(800) == 0 && d0 > (double) l && this.removeWhenFarAway(d0)) {
-                    this.discard();
+                    this.despawn(); // Monumenta
                 } else if (d0 < (double) l) {
                     this.noActionTime = 0;
                 }
diff --git a/src/main/java/net/minecraft/world/level/BaseSpawner.java b/src/main/java/net/minecraft/world/level/BaseSpawner.java
index 366c72a288c24732a65f6b7732e7d8e321bbb2cb..e6949cf4559073de85409ce7de41efb2e7e7ee09 100644
--- a/src/main/java/net/minecraft/world/level/BaseSpawner.java
+++ b/src/main/java/net/minecraft/world/level/BaseSpawner.java
@@ -46,6 +46,7 @@ public abstract class BaseSpawner {
     public int requiredPlayerRange = 16;
     public int spawnRange = 4;
     private int tickDelay = 0; // Paper
+    public BlockPos blockPos = null; // Monumenta
 
     public BaseSpawner() {}
 
@@ -187,6 +188,12 @@ public abstract class BaseSpawner {
                             // Spigot End
                         }
                         entity.spawnedViaMobSpawner = true; // Paper
+                        // Monumenta START
+                        if (!(entity instanceof net.minecraft.world.entity.FlyingMob || entity instanceof net.minecraft.world.entity.monster.Vex)) {
+                            entity.spawnerSpawnedBy = this;
+                        }
+                        // Monumenta END
+
                         entity.spawnReason = org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SPAWNER; // Paper
                         flag = true; // Paper
                         // Spigot Start
diff --git a/src/main/java/net/minecraft/world/level/block/entity/SpawnerBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/SpawnerBlockEntity.java
index 46a6f604ea14fe29d8a5f12dda4e533a4eb61637..5e49d7bc8ee06080a74a387b26eba70b849d85d1 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/SpawnerBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/SpawnerBlockEntity.java
@@ -33,6 +33,7 @@ public class SpawnerBlockEntity extends BlockEntity {
 
     public SpawnerBlockEntity(BlockPos pos, BlockState state) {
         super(BlockEntityType.MOB_SPAWNER, pos, state);
+        this.spawner.blockPos = pos; // Monumenta
     }
 
     @Override
