From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Timothy Southwick <NickNackGus@gmail.com>
Date: Thu, 27 Oct 2022 09:27:04 -0400
Subject: [PATCH] Monumenta - Freeze end gateway beacon


diff --git a/src/main/java/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
index c2bc747e9b9dd02971c13cb88e7aebeca8c0f2aa..b76afe9bcad9af8b3c74379bf5b7f2b203b232d9 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
@@ -56,6 +56,7 @@ public class TheEndGatewayBlockEntity extends TheEndPortalBlockEntity {
 
     public TheEndGatewayBlockEntity(BlockPos pos, BlockState state) {
         super(BlockEntityType.END_GATEWAY, pos, state);
+        this.age = SPAWN_TIME;
     }
 
     @Override
@@ -88,7 +89,7 @@ public class TheEndGatewayBlockEntity extends TheEndPortalBlockEntity {
     }
 
     public static void beamAnimationTick(Level world, BlockPos pos, BlockState state, TheEndGatewayBlockEntity blockEntity) {
-        ++blockEntity.age;
+        //++blockEntity.age;
         if (blockEntity.isCoolingDown()) {
             --blockEntity.teleportCooldown;
         }
@@ -99,7 +100,7 @@ public class TheEndGatewayBlockEntity extends TheEndPortalBlockEntity {
         boolean flag = blockEntity.isSpawning();
         boolean flag1 = blockEntity.isCoolingDown();
 
-        ++blockEntity.age;
+        //++blockEntity.age;
         if (flag1) {
             --blockEntity.teleportCooldown;
         } else {
