From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joel Ong <ongjiaming2001@gmail.com>
Date: Sat, 26 Nov 2022 15:37:35 +1100
Subject: [PATCH] Monumenta - Disable Bouncing Arrows


diff --git a/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java b/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
index 5d6d26cfe8f0ab68a3145214b3fc126ca7a71a66..fd2e16cb55980731122c43eb14124e90ac1a0a8f 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
@@ -451,6 +451,7 @@ public abstract class AbstractArrow extends Projectile {
             }
         } else {
             entity.setRemainingFireTicks(k);
+            /* Monumenta - Prevent Arrows from Bouncing off mobs.
             this.setDeltaMovement(this.getDeltaMovement().scale(-0.1D));
             this.setYRot(this.getYRot() + 180.0F);
             this.yRotO += 180.0F;
@@ -461,6 +462,10 @@ public abstract class AbstractArrow extends Projectile {
 
                 this.discard();
             }
+            */
+            if (this.getPierceLevel() <= 0) {
+                this.discard();
+            }
         }
 
     }
