From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joel Ong <ongjiaming2001@gmail.com>
Date: Sat, 26 Nov 2022 15:37:35 +1100
Subject: [PATCH] Monumenta - Disable Bouncing Arrows


diff --git a/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java b/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
index 53d0024daf6963ac4dab575666b0d6a74a39a958..fbc1b5cb7a7e5365e9897d5e9976bda3ffc5c702 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
@@ -459,6 +459,7 @@ public abstract class AbstractArrow extends Projectile {
             }
         } else {
             entity.setRemainingFireTicks(k);
+            /* Monumenta - Prevent Arrows from Bouncing off mobs.
             this.setDeltaMovement(this.getDeltaMovement().scale(-0.1D));
             this.setYRot(this.getYRot() + 180.0F);
             this.yRotO += 180.0F;
@@ -469,6 +470,10 @@ public abstract class AbstractArrow extends Projectile {
 
                 this.discard();
             }
+            */
+            if (this.getPierceLevel() <= 0) {
+                this.discard();
+            }
         }
 
     }
